include/mbedtls/ccm.h

```diff
@@ -288,6 +293,166 @@ int mbedtls_ccm_star_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
+ *                  any input with mbedtls_ccm_update().
+ * \param add       The buffer holding the additional data, or \c NULL
+ *                  if \p add_len is \c 0.
+ * \param add_len   The length of the additional data. If \c 0,
+ *                  \p add may be \c NULL.
+ *
+ * \return          \c 0 on success.
+ * \return          \#MBEDTLS_ERR_CCM_BAD_INPUT on failure:
+ *                  total input length too long.
+ */
+int mbedtls_ccm_update_ad( mbedtls_ccm_context *ctx,
+                           const unsigned char *add,
+                           size_t add_len );
+
+/**
+ * \brief           This function feeds an input buffer into an ongoing CCM
+ *                  encryption or decryption operation.
+ *
+ *                  You may call this function zero, one or more times
+ *                  to pass successive parts of the input: the plaintext to
+ *                  encrypt, or the ciphertext (not including the tag) to
+ *                  decrypt. After the last part of the input, call
+ *                  mbedtls_ccm_finish(). The lengths \p input_len of the
+ *                  data parts should eventually add up exactly to the total
+ *                  length of input data \c total_input_len passed to
+ *                  mbedtls_ccm_starts().
+ *
+ *                  This function may produce output in one of the following
+ *                  ways:
+ *                  - Immediate output: the output length is always equal
+ *                    to the input length.
+ *                  - Buffered output: but for the last part of input data,
+ *                    the output consists of a whole number of 16-byte blocks.
+ *                    If the total input length so far (not including
+ *                    associated data) is 16 \* *B* + *A* with *A* < 16 then
+ *                    the total output length is 16 \* *B*.
+ *                    For the last part of input data, the output length is
+ *                    equal to the input length plus the number of bytes (*A*)
+ *                    buffered in the previous call to the function (if any).
+ *                    The function uses the total length of input data
+ *                    \c total_input_len passed to mbedtls_ccm_starts()
+ *                    to detect the last part of input data.
+ *
+ *                  In particular:
+ *                  - It is always correct to call this function with
+ *                    \p output_size >= \p input_len + 15.
+ *                  - If \p input_len is a multiple of 16 for all the calls
+ *                    to this function during an operation (not necessary for
+ *                    the last one) then it is correct to use \p output_size
+ *                    =\p input_len.
+ *
+ * \note            This function is not implemented in Mbed TLS yet.
+ *
+ * \note            For decryption, the output buffer cannot be the same as
```

Comment:
I didn't get yet a comment about this but I didn't find any checks for that in ccm_auth_crypt(). Not sure what is supported regarding input and output buffer overlap though thus I kept it for additional views on that. @mpg @gilles-peskine-arm do you happen to know something about that?

Reply:
Support for buffer overlap is inconsistent across library modules, often undocumented, and I'm not convinced that all the documentation that is there is actually correct. For PSA we'll need to support arbitrary overlap to comply to the specification, but right now it's a known bug that our PSA crypto implementation does not support overlap.

Reply:
I don't think the 8 bytes constraint makes sense, by the way: I don't see any reason why 8 bytes would matter: pretty much all processing is one block at a time, and a block is 16 bytes for most ciphers. This looks like some bad copypasta from another place, maybe from very old code that only supported ciphers with 8-byte blocks. Fixing this elsewhere is out of scope, but maybe change 8 to 16 in the CCM module?

Reply:
Removed.